### ISSUE TITLE
[feature] Introduce concept of ABI groups for easier interoperability.

### ISSUE THREAD

**Issue by bmanga** [conan-io/conan#8074]
## Summary
Add an option to allow conan packages built with different - but ABI-compatible -
compilers to be used together.

## Motivation
Currently conan packages built with different compilers cannot be used in projects
using a different compiler, even if the generated binary is ABI-compatible.
While building all dependencies of a project with the same compiler can have some
clear benefits, not user scenarios warrant such constraint.
In fact, this fine-grained compiler requirement easily leads to bloat if the same
packages are used from projects using different compiler versions.
Additionally, we will lose out on pre-built binaries if the project switches to a compiler
version that is newer than the ones available in the conan-center-index CI.

Conan packages currently can work around this problem by overriding the `package_id`
function, but that is at the package creator's discretion and beyond control of the
package consumers.

## Proposal
* Introduce ABI versions for the various compiler versions/stdlib combinations.
* Add an additonal mechanism to opt-in/out on exact package compiler matches on a
per-project basis, with a default entry on the conan profiles.

- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @bmanga 

Thanks for your request. 

First, from our experience this claim:
> Currently conan packages built with different compilers cannot be used in projects
using a different compiler, even if the generated binary is ABI-compatible.

Is difficult to satisfy. Binary compatibility between the same compiler can be a challenge when changing versions, configuration (libstdc++, runtime, etc, etc). While we are aware than in some cases it is possible to achieve compatibility between compilers, that seems more an exception than the case. It is very possible that this compatibility can only be achieve for a subset of packages of a project (maybe pure C projects, while C++ can't achieve that binary compatibility)

The current approach to deal with this are:

- Detail the settings that needs to be used for each package, explicitly in profiles:
```
[settings]
compiler = gcc
pkg1:compiler = clang
pkg2:compiler = clang
```
This approach doesn't require any change in recipes, it can define easily which packages are built with what, and is also complete and easy to evolve, if want to change the used compiler for one package. The only problem is that it can be a bit verbose on the recipe, especially if the graph is huge and the compilers are uniformly spread, like 50% each.

- Use the ``compatible_packages`` features to fallback, explicitly in recipes :
```python
def package_id(self):
     if self.settings.compiler == "clang":
           pid = self.info.clone()
           pid.settings.compiler = "gcc" # fallback to gcc if no clang binary is available
           self.compatible_packages.append(pid)
```
This is the approach that has already been used to implement the intel compiler (icc) - gcc and icc - Visual compatibility. It has high level modeling with the ``compiler.base`` setting and helper functions to simplify ``package_id``.

If the fallback logic has to be applied in many recipes, the recommended approach would be to put it in a ``python_requires`` and reuse it from there.


To understand a bit more about the proposal, we would like to know some examples of these ABI groups, which compilers and which versions are considered compatible at the binary level? Under what assumptions? What other configuration like libcxx, runtime, etc needs to match? You can read in https://github.com/conan-io/conan/issues/7664, how some users say that the package_id is too strict, while others saying the opposite, it is not enough.

> Additionally, we will lose out on pre-built binaries if the project switches to a compiler
version that is newer than the ones available in the conan-center-index CI.

Yes, this can definitely happen. We will be adding new binaries for new mainstream compiler versions, but at some point, many users will need to build their own binaries, because there are no matching binaries available in ConanCenter. And in fact, many users do this, our stats show in fact that the majority of users do! There are many reasons, of course there are many companies that due to regulations or policies, they cannot use binaries they didn't build, so they use the recipes but build their own binaries. 

In any case, using your own Artifactory to cache your binaries is completely recommended, so you will not be affected by ConanCenter outages.  With that in place, building your own binaries is just one ``--build`` away, so my suggestion would be to embrace it, to be ready to build binaries for ConanCenter packages as well and put them in your Artifactory. My experience is that it cost less time to build the binaries for third parties with matching compilers, etc., than chasing very weird and difficult bugs in production because there is a subtle ABI incompatibility between different packages.

**Comment by puetzk** [comment]
A couple of big problems with this approach (in conan 1.x). Not blockers to the concept, they are likely fixable, but they keep it from being an attar active solution today.

1. Any use of `compatible_packages` defeats the local cache and requires a live connection to remotes, because the cache will never store a negative result "this package_id does not exist". Therefore, if the consumer's exact match is not going to be available (which is the only scenario in which compatible_packages will be useful), it will hit remotes every time, just to be sure, before it uses the compatible_package. Combine this with cmake-conan and an IDE that automatically reconfigures cmake when needed and this can get pretty annoying. Add remote work, an enterprise artifactory instance that requires VPN tunneling, and a locked-down corporate VPN with session timeouts, etc and it becomes nigh unbearable.

2. It can only declared by the producer recipe, not the consumer - so there's no way to do common things like `build_type=RelWithDebInfo` please, but `Release` will do if you don't have it). Or to express that release vs debug doesn't really matter so much with libstdc++/libc++, (unlike MSVCP, where `build_type=Debug` [has extra magic to imply `compiler.runtime=MDd`](https://github.com/conan-io/conan/blob/f230bf104842c160be6ea377497151fe1ea64c38/conans/client/settings_preprocessor.py#L49) and therefore impacts `_ITERATOR_DEBUG_LEVEL` (which does matter to ABI) as well as well as optimization and debug-symbol settings (which should logically be compatible_packages). One could envision something like that settings_preprocessor (maybe as a hook?) to apply knowlege of aparticular compiler's ABI rules and make better defaults, but there's no place to put such a thing right now except to copy&paste compatible_packages boilerplate into ~every recipe. Which is a bad idea given the drawbacks (1) of using any recipe that has compatible_packages right now.

**Comment by memsharded** [comment]
Conan 2.0 introduced the ``compatibility.py`` plugin that allows to define custom ABI groups at user control. Closing this as implemented in 2.0
